### PR TITLE
[Spec Decode] Adaptive verifier step-length for DFlash/PARD (D-Cut-style)

### PR DESCRIPTION
**Pull request by EanWang211123**
## Purpose

Add **adaptive verifier step-length** for parallel speculative decoding (DFlash, PARD / `draft_model` + `parallel_drafting`). The drafter still proposes the full `num_speculative_tokens` every step, but the verifier dynamically shortens the next step's `query_len` based on drafter softmax probabilities and a profiled ITL cost table.

### Motivation

- [PR #32374](https://github.com/vllm-project/vllm/pull/32374) shows that at high concurrency, truncated verifier shapes are more likely to run under PCG than FCG. The PCG–FCG latency gap at high batch sizes can be outweighed by the end-to-end gain from adaptive step-length pruning (less verify work on positions likely to be rejected).
- Core algorithm follows [D-Cut: Adaptive Verification Depth Pruning for Speculative Decoding](https://angelslim.readthedocs.io/zh-cn/latest/dcut.html): hardware-profiled cost table + draft-confidence prefix-product scores + batch-wide global top-K allocation to maximize `expected_accepted_tokens / verifier_ITL`.

### Runtime flow

```text
Step N
  execute_model start
    → read cached draft_len[req] from step N-1
    → truncate scheduled_spec_decode_tokens
      (draft_len=0 removes the spec entry for that request)
  verifier forward (truncated query lengths)
  propose_draft_token_ids
    → drafter still drafts full num_speculative_tokens
    → selected_probs[B, T] (softmax prob of each sampled draft token)
  sample_tokens (after _bookkeeping_sync)
    → choose_query_lens_discrete → cache draft_len[req] for step N+1

Step N+1
  execute_model start applies the cached draft_len …
```

`probs(N)` only affects `verifier(N+1)` (strict one-step lag).

### Core algorithm (D-Cut-style)

Verifier cost is profiled as `cost = ITL(batch_size, sum_query_len)`, where
`sum_query_len = B + Σ_i draft_lens[i]` (each sequence always contributes one
anchor token).

Per decode step, for active (post-prefill) sequences:

- **Input**: `probs[i][t]` — accept-probability proxy (softmax at the sampled
  draft token); `B` — full batch size; candidate `Q` — profiled `sum_query_len`
  keys for `bs_key` (ceil of `B` on the batch axis).
- **Marginal gain**: `m[i,t] = p[i,1] * … * p[i,t]`. Within each sequence `m`
  is non-increasing, so a global top-`S` pick satisfies the prefix constraint.
- **Score each candidate Q**: `S = Q - B`, `score = (B + sum(top-S gains)) / cost_table[(bs_key, Q)]`.
- **Output**: pick best `Q`, map top-`best_S` marginals to per-request
  `draft_len[i]`, cache for next-step truncation. Drafter still drafts full
  `num_speculative_tokens`; only verifier `query_len[i] = 1 + draft_len[i]` is
  adapted (0 to `num_speculative_tokens`).

### Implementation notes

- Probs transfer and CPU-side decision still add overhead; we avoid an extra
  blocking GPU sync on the critical path by piggybacking the D2H copy on the
  existing default-stream sync in `RejectionSampler.parse_output`.
- TP-safe: broadcast rank-0 `cost_table` after warmup profiling to avoid
  divergent `draft_lens`.

### Performance bottlenecks and outlook

**1. Low-concurrency overhead (~10%)**

At concurrency 1, enabling adaptive verify still shows a measurable throughput
drop vs static full-step verify (observed ~370 vs ~410 tok/s in internal runs;
exact numbers depend on model / GPU). The main suspected bottleneck is
`selected_probs` GPU→CPU transfer and the CPU-side marginal-gain decision — not
the extra softmax in the drafter itself (`needs_draft_probs` alone without the
controller is much cheaper).

Current mitigation (not zero-cost):
- Queue a **non-blocking** D2H copy of `selected_probs` to **pinned CPU memory**
  on the default CUDA stream in `_copy_draft_token_ids_to_cpu`.
- **Piggyback** on the existing default-stream sync in
  `RejectionSampler.parse_output` (inside `_bookkeeping_sync`), so we do not add
  a separate blocking `.cpu()` on the hot path.
- Run `choose_query_lens_discrete` at the end of `sample_tokens`, after
  bookkeeping, when the GPU is otherwise idle.

This removes the worst-case standalone D2H stall, but residual cost remains:
extra drafter logit/prob work, pinned-memory copy bandwidth, and NumPy decision
latency. **~10% low-concurrency regression is still expected today** and should
be called out in benchmarks.

**2. Complementary control with [PR #32374](https://github.com/vllm-project/vllm/pull/32374)**

This PR and PR #32374 target different levers and are intended to compose:

| | PR #32374 | This PR (adaptive verify) |
|---|---|---|
| When | **Prior** — at / around draft time | **Posterior** — after draft, before next verify |
| Signal | Draft-side scheduling / shape control | Draft confidence (`selected_probs`) |
| Effect | Better draft–verify shape alignment (e.g. PCG-friendly buckets) | Shorter verifier `query_len` on low-confidence tails |

Together: PR #32374 improves the drafting / graph-shape side; this PR prunes
verify depth from measured draft quality. At high concurrency the verify savings
can dominate; at low concurrency the combined story is still being evaluated.

**3. Async scheduling**

This change is developed and validated primarily on the **synchronous** DFlash
path (`use_gpu_toks`, drafter before `_bookkeeping_sync`). Async scheduling
(`async_scheduling=True`) uses a different draft / bookkeeping ordering and may
skip or defer `_copy_draft_token_ids_to_cpu`; the adaptive hooks are not yet
fully adapted to that pipeline. Follow-up work: relocate prob capture / consume
to the async-specific draft sync points without breaking the one-step-lag
correctness guarantee (`probs(N)` → `verifier(N+1)`).

### New modules

| File | Summary |
|------|---------|
| `vllm/v1/spec_decode/verify_adaptive_config.py` | JSON-backed config dataclass: profiling grid (batch / query_len axes), `warmup_seq_lens`, measurement iters, validation. |
| `vllm/v1/spec_decode/verify_adaptive_controller.py` | Core controller: `profile_cost_table`, NumPy `choose_query_lens_discrete`, per-request `draft_lens` cache, TP broadcast of cost table. |
| `vllm/v1/spec_decode/verify_adaptive.md` | Feature doc: enablement, runtime flow, cost model, code map. |

### Config & CLI

| File | Summary |
|------|---------|
| `vllm/config/speculative.py` | Add `speculative_adaptive_verify_config` field and `supports_adaptive_verify()` (DFlash or parallel `draft_model` only). |
| `vllm/engine/arg_utils.py` | Add `--speculative-adaptive-verify-config` CLI flag and speculative-config JSON key wiring. |

### Drafter: draft-token probability collection

| File | Summary |
|------|---------|
| `vllm/v1/spec_decode/llm_base_proposer.py` | Add `needs_draft_probs`, `_last_selected_probs`, `take_last_selected_probs()`, `_gather_selected_probs()` (logsumexp, no full-vocab softmax). Extend `_sample_draft_tokens` to return `selected_probs` for parallel and serial draft paths. |

### Runner integration

| File | Summary |
|------|---------|
| `vllm/v1/worker/gpu_model_runner.py` | Init controller + pinned CPU buffer; truncate `scheduled_spec_decode_tokens` at `execute_model` start; non-blocking D2H in `_copy_draft_token_ids_to_cpu` (residual CPU/numpy cost remains); consume probs in `sample_tokens` after `_bookkeeping_sync`; `invalidate` on request finish; `_adaptive_profile_run` + `profile_adaptive_cost`. |
| `vllm/v1/worker/gpu_worker.py` | Call `profile_adaptive_cost()` at end of `compile_or_warm_up_model` (after CUDA-graph / JIT warmup). |

## Test Plan

- [x] **Qwen3-8B, 2× RTX 4090, TP=2**: compare AR baseline vs DFlash/PARD static
  (full verify) vs dynamic (adaptive verify) at concurrency 4 / 16 / 32 / 48.
- [x] Enable adaptive verify on DFlash and confirm `adaptive:` debug logs appear
  after warmup.
- [x] TP=2 sustained load: no NCCL / RPC timeout observed.
- [x ] Verify correctness lag: step-N `selected_probs` only changes step-(N+1)
  verifier length (inspect per-step `draft_lens` logs).

Example serve command (DFlash + adaptive verify):

```bash
vllm serve Qwen/Qwen3-8B \
  --tensor-parallel-size 2 \
  --spec-method dflash \
  --spec-model <dflash_draft_model> \
  --spec-tokens 15 \
  --speculative-adaptive-verify-config /path/to/verify_adaptive_config.json
```

## Test Result

**Setup**: Qwen3-8B, 2× RTX 4090, TP=2, `output_throughput` (tok/s).

| Concurrency | 4 | 16 | 32 | 48 |
|-------------|---:|---:|---:|---:|
| **DFlash — AR baseline** | 331.68 | 967.21 | 1794.32 | 1612.29 |
| **DFlash — static** (full verify) | 1057.60 | 1896.82 | 2347.33 | 2451.30 |
| **DFlash — dynamic** (adaptive verify) | 926.87 | 2141.59 | 2784.74 | 2638.20 |

| Concurrency | 4 | 16 | 32 | 48 |
|-------------|---:|---:|---:|---:|
| **PARD — AR baseline** | 335.68 | 883.11 | 1479.43 | 1703.14 |
| **PARD — static** (full verify) | 891.91 | 1589.51 | 2001.02 | 2197.50 |
| **PARD — dynamic** (adaptive verify) | 643.67 | 1454.21 | 2517.18 | 2715.72 |

**Observations**

- **Low concurrency (bs=4)**: dynamic is below static for both DFlash (−12%) and
  PARD (−28%), consistent with the residual probs D2H / CPU-decision overhead
  discussed above.
- **Mid–high concurrency (bs≥16)**: dynamic overtakes static; largest gain at
  bs=32 (DFlash +19%, PARD +26% vs static). At bs=48, dynamic still leads static
  on both methods.
- **vs AR baseline** (DFlash): at bs=32 dynamic **1.55×** / static **1.31×**; at
  bs=48 dynamic **1.64×** / static **1.52×**. Adaptive verify extends the
  high-concurrency advantage of parallel speculative decoding.

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model (`verify_adaptive.md` added).
</details>

### CODE REVIEW

**Comment by benchislett** [comment]
As discussed here, the surface area of this feature is too significant to justify the marginal and hardware-specific gains https://github.com/vllm-project/vllm/pull/35301

**Comment by EanWang211123** [comment]
@benchislett 

Thank you very much for the reply! I understand the concern that this feature currently introduces a relatively large amount of change and may be difficult to maintain. However, I would like to add a few points for consideration:

1. **Better hardware adaptivity.**
   Both #35301 and #43522 are based on setting a threshold over the draft probability to decide where to truncate. However, if the hardware changes, this threshold will likely need to be retuned. In contrast, a throughput-aware approach considers both the additional overhead of speculative decoding under different workloads and the expected benefit of drafting/verifying different numbers of steps. This makes it possible to adapt to different hardware platforms and workloads automatically. Therefore, I do not think this is only a hardware-specific optimization. In my tests, I observed around a 20% throughput improvement under medium-to-high concurrency on a 2x RTX 4090 setup, and around a 40–50% throughput improvement on H20.

2. **Per-sequence isolation within the same batch.**
   As mentioned in #35301, if we use batch size as the observation granularity and force all sequences in the batch to share the same dynamic step length, we cannot eliminate the differences between individual sequences. A better approach is to make decisions at the per-sequence level, since even within the same batch, some sequences may have longer query lengths while others may be shorter.

3. **Ease of use in production.**
   Another reason why I believe this feature is valuable is that, when serving many different customers, it is often necessary to manually tune hyperparameters such as the number of draft steps. With this feature, the system can fit toward a better operating point automatically without requiring manual intervention for each customer's parameters. It can also adapt the verify length and draft length at runtime as the online workload changes. As far as I can tell, the approaches in the above PRs, as well as #32374, still require users to manually configure thresholds or other parameters, so they do not fully close the loop from an engineering perspective.

4. **System intrusiveness.**
   Although I agree that the current code structure still has a lot of room for improvement, I think the actual algorithmic intrusion is mainly limited to adding some operations after drafting and before verification. This may introduce some synchronization/blocking overhead, which also seems to be a concern in other PRs, and I am currently trying to address it.

5. **Algorithmic rationale.**
   Based on my experiments, the optimization brought by speculative decoding is roughly determined by the average accepted length after drafting several steps divided by the ITL overhead under a specific workload. The latter can be estimated through a dummy-run at startup, while the former can be estimated using the draft probability. Therefore, I believe the algorithm is reasonably well motivated.

image


Overall, I still believe that the optimization potential and engineering stability of dynamic step length have not yet reached their upper bound. I also think that hardware-dependent overhead is a factor that should not be ignored. Although this PR introduces relatively large changes compared with the current code structure, I hope this approach can still be considered as a candidate for future dynamic step-length development, especially after some more fundamental dynamic step-length solutions are merged.I would also be very happy to discuss this further and learn how the community thinks about dynamic step length for speculative decoding in general.

**Comment by ooolmk** [comment]
Since using draft confidence as the cutoff criterion cannot avoid slowdowns under low concurrency, would it be more straightforward and effective to cut verification tokens directly based on the concurrency level instead?

**Comment by benchislett** [comment]
This feature is now supported! @ooolmk 

https://docs.vllm.ai/en/latest/features/speculative_decoding/dynamic_speculative_decoding

**Comment by EanWang211123** [comment]
@benchislett  @ooolmk  Yes, the main branch already supports batch-size-aware + DFlash dynamic drafting. However, while I was experimenting with other parallel drafting speculative decoding algorithms (such as PARD and P-EAGLE), I found a small issue that needs to be fixed.

[https://github.com/vllm-project/vllm/pull/45695](https://github.com/vllm-project/vllm/pull/45695)
